### ISSUE TITLE
Properly free NaClApp structs

### ISSUE THREAD

**Issue by rennergade** [Lind-Project/lind_project#215]
My journey into deconstructing cages revealed that we're not properly freeing the NaClApp structs after we teardown the threads. I added a Dtor function for the struct but adding it with our current setup has proven somewhat hazardous because of concurrency issues.

**Comment by rennergade** [comment]
Instead of packing this into my signal commit I'm going to save this Dtor prototype here

```
// Dtor function for the nap structure
// TODO: implement in system
void NaClAppDtor(struct NaClApp *nap) {

  NaClLog(3, "Deconstructing nap\n");
  NaClLog(3, "Freeing Address Space\n");
  NaClAddrSpaceFree(nap);
  NaClLog(3, "Tearing down mutexes\n");
  NaClFastMutexDtor(&nap->desc_mu);
  NaClMutexDtor(&nap->threads_mu);
  // NaClDescUnref(nap->name_service_conn_cap);
  // NaClRefCountUnref((struct NaClRefCount *) nap->name_service);
  NaClCondVarDtor(&nap->cv);
  NaClMutexDtor(&nap->mu);
  NaClMutexDtor(&nap->children_mu);
  NaClMutexDtor(&nap->dynamic_load_mutex);
  free(nap->effp);
  NaClLog(3, "Deleting io regions\n");
  NaClIntervalMultisetDelete(nap->mem_io_regions);
  nap->mem_io_regions = NULL;
  // NaClVmmapDtor(&nap->mem_map);
  NaClLog(3, "Tearing down dyn arrays\n");
  DynArrayDtor(&nap->children);
  DynArrayDtor(&nap->desc_tbl);
  DynArrayDtor(&nap->threads);
  free(nap->cpu_features);
  free((void*) nap->clean_environ);

  NaClLog(3, "Freeing nap\n");
  NaClAlignedFree(nap);
}
```

One issue is that the Vmmap/DescUnref calls will trigger close() calls to rustposix so we need to disable that.

**Comment by rennergade** [comment]
I was able to fix this up by adding a cv to signal the cage has properly exited before cleanup. This also exposed problems with when a program creates orphan cages, so I fixed that as well. To properly Dtor the vmmap, I changed NaClSysExit to close all open fd's from NaCl before we exit the cage in safeposix and call the vmmap dtor.

This should get all of the nap destruction working properly. Will merge this with the cage exiting branch.
